src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexCompiler.cs

```diff
@@ -3143,17 +3151,19 @@ void EmitSingleCharLoop(RegexNode node, RegexNode? subsequent = null, bool emitL
                     BleFar(doneLabel);
                 }
 
-                if (!rtl && subsequent?.FindStartingLiteral() is ValueTuple<char, string?, string?> literal)
+                if (!rtl && subsequent?.FindStartingLiteral() is ValueTuple<char, string?, string?, bool> literal)
                 {
                     // endingPos = inputSpan.Slice(startingPos, Math.Min(inputSpan.Length, endingPos + literal.Length - 1) - startingPos).LastIndexOf(literal);
                     // if (endingPos < 0)
                     // {
                     //     goto doneLabel;
                     // }
+                    bool negated = literal.Item4;
                     Ldloca(inputSpan);
                     Ldloc(startingPos);
                     if (literal.Item2 is not null)
                     {
+                        Debug.Assert(!negated, "strings should not be negated");
```

Comment:
NIT: I didn't see the Debug.Assert in the source generator.

Reply:
Yeah, I didn't add one there because in the source generator the entire thing is nicely set up as a switch expression, and asserts don't play nicely with such expressions.